rfcs/convergence/context-values.md

```diff
@@ -0,0 +1,227 @@
++function useRenderBaz(state) {
+  const { open, inline } = state;
+  // ✅ now we can use hooks
+  const contextValue = React.useMemo(() => ({ inline, open }), [inline, open]);
+
+  // ...
+}
+```
+
+It's one of the easiest options, but it could cause issues in future as it breaks separation of concerns. For example, other hooks can be also called now in `useRender`:
+
+```tsx
+function useRenderBaz(state) {
+  const { open, inline } = state;
+
+  // 💣 other hooks can be also declared there
+  React.useEffect(() => {
+    // ...
+  }, [open]);
+}
+```
+
+- 👎 breaks separation of concerns as allows to use other hooks in render functions
+
+### Option 2: Use `useContextSelector()` everywhere
+
+For `context-selector` memoization is not required as it will happen in consumers. In this case we don't need to transform `render` functions to React hooks.
+
+```tsx
+function renderBaz(state) {
+  const { open, inline } = state;
+
+  return (
+    <SampleContext.Provider value={{ inline, open }}>
+      <div />
+    </SampleContext.Provider>
+  );
+}
+
+// ---
+
+function useBazItem() {
+  // ✅ only if "open"/"inline" will change a rerender will be triggered
+  const open = useContextSelector(SampleContext, value => value.open);
+  const inline = useContextSelector(SampleContext, value => value.inline);
+}
+```
+
+- 👍 Allows to keep existing API design
+
+- 👎 Even if it's enough performant option, it will be anyway slower than `React.useContext()` as we are evaluating additional code during render cycles
+- 👎 (_potentially_) This implementation is tested with React's Concurrent mode, but still could cause issues with React 18
+- 👎 Sooner or later `useContextSelector` will be implemented in React's core ([facebook/react#20646](https://github.com/facebook/react/pull/20646)) and it could have different API than our implementation
+
+### Option 3: Modify render functions to accept context value
+
+Another option is to modify a signature of render functions to include context values as the second param and compute them inside component.
+
+```diff
+function FooComponent() {
+  const state = useFooState()
+
++ const barContextValue = React.useMemo(/* ... */)
++ const bazContextValue = React.useMemo(/* ... */)
+
+- return renderFoo(state)
++ return renderFoo(state, { bar: barContextValue, baz: bazContextValue })
+}
+
+// ---
+
+-function renderFoo(state) {
+-  return <SampleContext.Provider value={state.bazContextValue} />
++function renderFoo(state, contextValues) {
++  return <SampleContext.Provider value={contextValues.baz} />
+}
+```
+
+To improve experience for customers and ourselves we can extract creation of values in a separate hook:
+
+```tsx
+function useFooContextValues() {
+  const barContextValue = React.useMemo(/* ... */);
+  const bazContextValue = React.useMemo(/* ... */);
+
+  return { bar: bazContextValue, baz: bazContextValue };
+}
+
+function FooComponent() {
+  const state = useFooState();
+  const contextValues = useFooContextValues();
```

Comment:
I think we also need to use state here, since sometimes context can be derived from state

```suggestion
  const contextValues = useFooContextValues(state);
```